src/lsptoolshost/restore.ts

```diff
@@ -49,22 +60,25 @@ async function chooseProjectAndRestore(
         return;
     }
 
-    await restore(languageServer, restoreChannel, pickedItem.description);
+    await restore(languageServer, restoreChannel, [pickedItem.description!], true);
 }
 
-async function restore(
+export async function restore(
     languageServer: RoslynLanguageServer,
     restoreChannel: vscode.OutputChannel,
-    projectFile?: string
+    projectFiles: string[],
+    showOutput: boolean
 ): Promise<void> {
     if (_restoreInProgress) {
         vscode.window.showErrorMessage(vscode.l10n.t('Restore already in progress'));
```

Comment:
Is there a possibility we can now hit this if automatic restore is on? I'm thinking something like the user invokes restore, it starts to restore projects, and tries (but fails) to fully restore one project, call it A. It's still going and marching to restore other projects in the solution, but we saw the restore of A, and then trigger a reload which triggers a check of A's dependencies. We see it's missing, and so we trigger a restore of A _again_, but the restore of the rest of the solution is still going on...

Should we be queueing these instead in some way? That might also make sense if somebody is invoking two project restores one after the other in rapid succession.

Reply:
This should be safe on the reload side because the restores on the client will block the project load queue until they're done.  So the reload won't be processed until the rest of the projects finish restoring (I actually hit this issue with the first iteration with notifications)

This should only get hit if someone triggers a manual restore while an automatic one is in progress

Reply:
> This should only get hit if someone triggers a manual restore while an automatic one is in progress

Yeah that was the case that I had in mind.

Reply:
Yeah it is possible to hit this with a manual restore.  I think for now though we don't need a queue - it should be a relatively uncommon case.  If its not and ends up being a bad experience for people we can come back and add it.